FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestProvider.m

```diff
@@ -172,6 +174,29 @@ - (void)getTokenWithCompletion:(void (^)(FIRAppCheckToken *_Nullable, NSError *_
 }
 
 - (FBLPromise<FIRAppCheckToken *> *)getToken {
+  return [FBLPromise onQueue:self.queue
+                          do:^id _Nullable {
+                            if (self.ongoingGetTokenOperation == nil) {
+                              // Kick off a new handshake sequence only when there is not an ongoing
+                              // handshake to avoid race conditions.
+                              self.ongoingGetTokenOperation =
+                                  [self createGetTokenSequencePromise]
+
+                                      // Release the ongoing operation promise on completion.
+                                      .then(^FIRAppCheckToken *(FIRAppCheckToken *token) {
+                                        self.ongoingGetTokenOperation = nil;
+                                        return token;
+                                      })
+                                      .recover(^NSError *(NSError *error) {
+                                        self.ongoingGetTokenOperation = nil;
+                                        return error;
+                                      });
+                            }
+                            return self.ongoingGetTokenOperation;
```

Comment:
IIUC, we want to return a promise that resolves to a token (or error) and nils out the `ongoingGetTokenOperation` that it was being stored in when it finally resolves? 

While understanding this, I thought of something I'd like your thoughts on:
Could there be a race condition that emerges when the `.then` or `.recover` blocks execute fast enough where its's synchronous (particularly the `.recover` error case)? Because then `ongoingGetTokenOperation` may persist with a value and prevent future calls to `getToken` from spawning a new `getTokenSequencePromise`

Reply:
Good question.

 Yes, the intention is to return a promise that nils out the `ongoingGetTokenOperation` on completion. 

Currently `ongoingGetTokenOperation` stores such a promise (includes `then` and `recovery` steps). So the intended sequence of events is:
1. the promise created and assigned to `ongoingGetTokenOperation`
2. new promises bound to `ongoingGetTokenOperation` are created in `getTokenWithCompletion:` method to invoke the completion blocks
3. the promise created by `[self createGetTokenSequencePromise]` is either fulfilled or rejected
4. `ongoingGetTokenOperation` is nilled out
  - if another `getTokenWithCompletion:`  method is called at this point the sequence will start from step 1.
5. promises created in `getTokenWithCompletion:`  are fulfilled or rejected correspondingly

The sequence looks correct to me and I cannot see a possible race condition here. Can you see anything missing/off?

Reply:
Thank you for the detailed explanation. I think it looks good. I'm still trying to understand the ins and outs of how to closely examine promises.

Here I was trying to make sure that `ongoingGetTokenOperation ` is never returned from `getToken:` in a resolved state, because that would prevent other calls to `getToken:` from doing anything.

I think my original doubts were for when the following occurs:
1. `getTokenWithCompletion:` is called but app attest is not supported, so an `attestState.appAttestUnsupportedError` is returned from `createGetTokenSequencePromise:`
2. If this evaluates fast enough (because of the lack of network calls) that it's synchronous, then might the `.recover` block return an error to  `ongoingGetTokenOperation` **before** the process returns from `getToken:`. This means that the `ongoingGetTokenOperation` is returned from `getToken:` in a resolved state which would be bad for future called of `getToken:`.
3. Enabling app attest support and then calling `getTokenWithCompletion` results in nothing because the  `ongoingGetTokenOperation` has persisted.

After taking a closer took, I think we are safe from this because  `attestationState:` executes on another thread so it should prevent the `ongoingGetTokenOperation` from resolving before being returned from `getToken`

Reply:
Ah, I see what you mean. Yeah, currently `then` or `recover` blocks will always be executed after the promise is returned. Thank you for the discussion!